# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [3]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Fname             999 non-null    object
 1   Lname             999 non-null    object
 2   Position          999 non-null    object
 3   Gender            999 non-null    object
 4   Hours Worked      999 non-null    int64 
 5   SalesTraining     999 non-null    object
 6   Years Experience  999 non-null    int64 
 7   Cars Sold         999 non-null    int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [5]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [6]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [7]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [8]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

In [9]:
poke = pd.read_csv("pokemon.csv")

poke.head()

Num        Name Type 1  Type 2  Total  HP  Attack  Defense  SpAtk  SpDef  \
0    1   Bulbasaur  Grass  Poison    318  45      49       49     65     65   
1    2     Ivysaur  Grass  Poison    405  60      62       63     80     80   
2    3    Venusaur  Grass  Poison    525  80      82       83    100    100   
3    4  Charmander   Fire     NaN    309  39      52       43     60     50   
4    5  Charmeleon   Fire     NaN    405  58      64       58     80     65   

   Speed  Stage  Legendary  
0     45      1      False  
1     60      2      False  
2     80      3      False  
3     65      1      False  
4     80      2      False

In [10]:
poke["Type 1"].value_counts()

Water       28
Normal      22
Poison      14
Grass       12
Fire        12
Bug         12
Electric     9
Rock         9
Ground       8
Psychic      8
Fighting     7
Ghost        3
Dragon       3
Fairy        2
Ice          2
Name: Type 1, dtype: int64

In [11]:
poke["Type 1"].value_counts(normalize = True)

Water       0.185430
Normal      0.145695
Poison      0.092715
Grass       0.079470
Fire        0.079470
Bug         0.079470
Electric    0.059603
Rock        0.059603
Ground      0.052980
Psychic     0.052980
Fighting    0.046358
Ghost       0.019868
Dragon      0.019868
Fairy       0.013245
Ice         0.013245
Name: Type 1, dtype: float64

In [12]:
pd.crosstab(poke["Legendary"], poke["Stage"], margins = True, normalize = False)

Stage       1   2   3  All
Legendary                 
False      75  56  16  147
True        4   0   0    4
All        79  56  16  151

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [13]:
df.head(2)

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6

In [14]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [15]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [16]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
Name: Hours Worked, dtype: int64

In [17]:
df.head(20)

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0     Jackie    Jackson  Trainee      F            32             Y   
1       Mary  Patterson  Trainee      F            43             N   
2      Tanya      Adams  Trainee      F            28             Y   
3      Tanya  Henderson  Trainee      F            24             Y   
4     Walter   Franklin  Trainee      M            25             Y   
5   Victoria   Ettienne  Trainee      F            29             Y   
6     Martin   Ettienne  Trainee      M            42             N   
7      Peter   Michaels  Trainee      M            43             N   
8     Denise  Henderson  Trainee      F            38             N   
9     Walter      Moore  Trainee      M            32             Y   
10       Sam   Michaels  Trainee      M            45             Y   
11      Jada   Walters   Trainee      F            24             Y   
12      Jack   Ettienne  Trainee      M            21             Y   
13     Tanya     Carter  Trainee      F            40             Y   
14       Tom    Mulgrew  Trainee      M            29             Y   
15     Betty     Looner  Trainee      F            41             Y   
16      John       Rose  Trainee      M            42             Y   
17     Roger       Sapp  Trainee      M            41             Y   
18     Karen  Samuelson  Trainee      F            41             Y   
19     india   Franklin  Trainee      F            36             Y   

    Years Experience  Cars Sold  
0                  1          3  
1                  1          6  
2                  1          2  
3                  1          5  
4                  1          4  
5                  1          2  
6                  1          6  
7                  1          6  
8                  1          5  
9                  1          3  
10                 1          3  
11                 1          4  
12                 1          2  
13                 1          4  
14                 1          5  
15                 1          6  
16                 1          2  
17                 1          3  
18                 1          2  
19                 1          2

In [18]:
df["Cars Sold"].value_counts()

5    206
4    202
3    194
2    172
6    156
1     44
7     25
Name: Cars Sold, dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

In [19]:
poke["Attack"].mean()

72.54966887417218

In [20]:
poke["Defense"].median()

65.0

In [21]:
poke["Speed"].mode()

0    70
1    90
Name: Speed, dtype: int64

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [22]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [23]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [24]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [25]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 


#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [26]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [27]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

In [28]:
# Calculate the range of values in HP column

hp_range = poke['HP'].max() - poke['HP'].min()

print(hp_range)

240


In [29]:
# Mean and the standard deviation for the special attack (SpAtk) column
poke["SpAtk"].describe()

count    151.000000
mean      67.139073
std       28.534199
min       15.000000
25%       45.000000
50%       65.000000
75%       87.500000
max      154.000000
Name: SpAtk, dtype: float64

In [30]:
# Mean for SpAtk is 67.139073 and the standard deviation is 28.534199

In [31]:
# For the Attack column, determine which value falls at the 80th percentile? 100.0
# Which value falls at the 15th percentile? 45.0
# The mean for the 80th percentile is 131.80000
# The mean for the 15th percentile is 27.410611

poke['Attack'].quantile(.15) 


45.0

In [32]:
poke['Attack'].quantile(.85)

100.0

In [33]:
poke.describe().quantile(.85)

Num        149.125000
Total      485.853974
HP         147.450000
Attack     131.800000
Defense    147.650000
SpAtk      147.825000
SpDef      122.750000
Speed      137.500000
Stage        2.950000
Name: 0.85, dtype: float64

In [34]:
poke.describe().quantile(.15)

Num         38.761702
Total      153.200000
HP          29.410611
Attack      27.816354
Defense     28.070869
SpAtk       29.357489
SpDef       25.438029
Speed       27.734536
Stage        1.000000
Name: 0.15, dtype: float64

In [35]:
# Calculate the second quartile (or the median) for the special defense (SpDef) column

# Calculate the median of the same column using the method shown above. 
# Are the values the same? 

In [36]:
poke["SpDef"].median()

65.0

In [37]:
poke.describe().quantile(.50)

Num         76.000000
Total      362.500000
HP          62.105960
Attack      71.274834
Defense     66.612583
SpAtk       66.069536
SpDef       65.509934
Speed       69.466887
Stage        1.291391
Name: 0.5, dtype: float64

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [38]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [39]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [40]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)
 
dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [41]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [42]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [43]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [44]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [45]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [46]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [47]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [48]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [49]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [50]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [51]:
df = pd.read_excel("babies.xlsx")
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


In [53]:
df.describe()

bwt    gestation       parity          age       height  \
count  1236.000000  1223.000000  1236.000000  1234.000000  1214.000000   
mean    119.576861   279.338512     0.254854    27.255267    64.047776   
std      18.236452    16.027693     0.435956     5.781405     2.533409   
min      55.000000   148.000000     0.000000    15.000000    53.000000   
25%     108.750000   272.000000     0.000000    23.000000    62.000000   
50%     120.000000   280.000000     0.000000    26.000000    64.000000   
75%     131.000000   288.000000     1.000000    31.000000    66.000000   
max     176.000000   353.000000     1.000000    45.000000    72.000000   

            weight        smoke  
count  1200.000000  1226.000000  
mean    131.411667     0.394780  
std      53.687343     0.489003  
min      27.000000     0.000000  
25%     114.000000     0.000000  
50%     125.000000     0.000000  
75%     140.000000     1.000000  
max    1500.000000     1.000000

3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [54]:
df.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [55]:
# The two qualitative variables in this dataset are smoke and parity

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [56]:
# replace a specific value in a specific column

df.replace(1, "Yes", inplace = True)
df.replace(0, "No", inplace = True)

# check changes

df

bwt  gestation parity   age  height  weight smoke
0     120      284.0     No  27.0    62.0   100.0    No
1     113      282.0     No  33.0    64.0   135.0    No
2     128      279.0     No  28.0    64.0    77.0   Yes
3     123        NaN     No  36.0    69.0   190.0    No
4     108      282.0     No  23.0    67.0   125.0   Yes
...   ...        ...    ...   ...     ...     ...   ...
1231  113      275.0    Yes  27.0    60.0   100.0    No
1232  128      265.0     No  24.0    67.0   120.0    No
1233  130      291.0     No  30.0    65.0    67.0   Yes
1234  125      281.0    Yes  21.0    65.0   110.0    No
1235  117      297.0     No  38.0    65.0   129.0    No

[1236 rows x 7 columns]

6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [57]:
df.isnull()

bwt  gestation  parity    age  height  weight  smoke
0     False      False   False  False   False   False  False
1     False      False   False  False   False   False  False
2     False      False   False  False   False   False  False
3     False       True   False  False   False   False  False
4     False      False   False  False   False   False  False
...     ...        ...     ...    ...     ...     ...    ...
1231  False      False   False  False   False   False  False
1232  False      False   False  False   False   False  False
1233  False      False   False  False   False   False  False
1234  False      False   False  False   False   False  False
1235  False      False   False  False   False   False  False

[1236 rows x 7 columns]

In [58]:
df.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   object 
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   object 
dtypes: float64(4), int64(1), object(2)
memory usage: 67.7+ KB


7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [60]:
## bwt 115 has the highst Frequencies of 36 

df["bwt"].value_counts()


115    36
117    35
129    34
123    33
120    31
       ..
166     1
69      1
161     1
176     1
167     1
Name: bwt, Length: 107, dtype: int64

In [61]:

## Relative Frequencies 

df["bwt"].value_counts(normalize=True)

115    0.029126
117    0.028317
129    0.027508
123    0.026699
120    0.025081
         ...   
166    0.000809
69     0.000809
161    0.000809
176    0.000809
167    0.000809
Name: bwt, Length: 107, dtype: float64

In [62]:
## gestation 282.0 has the highst Frequencies of 48

df["gestation"].value_counts()

282.0    48
278.0    44
280.0    44
277.0    42
281.0    42
         ..
310.0     1
320.0     1
239.0     1
243.0     1
321.0     1
Name: gestation, Length: 106, dtype: int64

In [63]:

## Relative Frequencies 

df["gestation"].value_counts(normalize=True)

282.0    0.039248
278.0    0.035977
280.0    0.035977
277.0    0.034342
281.0    0.034342
           ...   
310.0    0.000818
320.0    0.000818
239.0    0.000818
243.0    0.000818
321.0    0.000818
Name: gestation, Length: 106, dtype: float64

In [64]:
# parity no has the highst Frequencies of 921

df["parity"].value_counts()

No     921
Yes    315
Name: parity, dtype: int64

In [65]:

## Relative Frequencies 

df["parity"].value_counts(normalize=True)

No     0.745146
Yes    0.254854
Name: parity, dtype: float64

In [66]:
## age 23 has the highest Frequencies of 93 

df["age"].value_counts()

23.0    93
26.0    90
24.0    86
27.0    85
22.0    79
25.0    77
28.0    70
21.0    67
29.0    66
30.0    63
20.0    58
19.0    53
31.0    46
33.0    42
32.0    39
34.0    33
35.0    30
37.0    29
36.0    26
39.0    24
38.0    18
18.0    15
41.0    14
40.0    11
17.0     7
43.0     6
42.0     4
15.0     1
44.0     1
45.0     1
Name: age, dtype: int64

In [67]:

## Relative Frequencies 

df["age"].value_counts(normalize=True)

23.0    0.075365
26.0    0.072934
24.0    0.069692
27.0    0.068882
22.0    0.064019
25.0    0.062399
28.0    0.056726
21.0    0.054295
29.0    0.053485
30.0    0.051053
20.0    0.047002
19.0    0.042950
31.0    0.037277
33.0    0.034036
32.0    0.031605
34.0    0.026742
35.0    0.024311
37.0    0.023501
36.0    0.021070
39.0    0.019449
38.0    0.014587
18.0    0.012156
41.0    0.011345
40.0    0.008914
17.0    0.005673
43.0    0.004862
42.0    0.003241
15.0    0.000810
44.0    0.000810
45.0    0.000810
Name: age, dtype: float64

In [68]:
#  height 64 has the highst Frequencies of 183

df["height"].value_counts()

64.0    183
65.0    182
63.0    166
66.0    153
62.0    131
67.0    105
61.0    105
60.0     55
68.0     54
59.0     26
69.0     20
70.0     13
58.0     10
71.0      6
56.0      1
72.0      1
53.0      1
57.0      1
54.0      1
Name: height, dtype: int64

In [69]:

## Relative Frequencies 

df["height"].value_counts(normalize=True)

64.0    0.150741
65.0    0.149918
63.0    0.136738
66.0    0.126030
62.0    0.107908
67.0    0.086491
61.0    0.086491
60.0    0.045305
68.0    0.044481
59.0    0.021417
69.0    0.016474
70.0    0.010708
58.0    0.008237
71.0    0.004942
56.0    0.000824
72.0    0.000824
53.0    0.000824
57.0    0.000824
54.0    0.000824
Name: height, dtype: float64

In [70]:
## weight 130.0 has the highest Frequencies of 77

df["weight"].value_counts()

130.0    77
125.0    68
110.0    64
135.0    62
120.0    54
         ..
845.0     1
399.0     1
197.0     1
171.0     1
67.0      1
Name: weight, Length: 116, dtype: int64

In [71]:

## Relative Frequencies 

df["weight"].value_counts(normalize=True)

130.0    0.064167
125.0    0.056667
110.0    0.053333
135.0    0.051667
120.0    0.045000
           ...   
845.0    0.000833
399.0    0.000833
197.0    0.000833
171.0    0.000833
67.0     0.000833
Name: weight, Length: 116, dtype: float64

In [72]:
## smoke no has the highest Frequencies 742

df["smoke"].value_counts()

No     742
Yes    484
Name: smoke, dtype: int64

In [73]:

## Relative Frequencies 

df["smoke"].value_counts(normalize=True)

No     0.60522
Yes    0.39478
Name: smoke, dtype: float64

In [74]:
# How many mothers are smokers who have been pregnant before? 121

pd.crosstab(df["smoke"], df["parity"], margins = True)


parity   No  Yes   All
smoke                 
No      548  194   742
Yes     363  121   484
All     911  315  1226